lib/handler/pool.ex

```diff
@@ -0,0 +1,57 @@
+defmodule Handler.Pool do
+  defstruct max_workers: 100,
+            max_memory_bytes: 10 * 1024 * 1024 * 1024,
+            name: nil
+
+  alias __MODULE__
+  alias Handler.Pool.State
+  use GenServer
+
+  def start_link(%Pool{} = config) do
+    GenServer.start_link(Pool, config)
+  end
+
+  def attempt_work(pool, fun, opts) do
+    GenServer.call(pool, {:attempt, fun, opts})
+  end
+
+  def init(%Pool{} = pool) do
+    register_name(pool)
+    state = %State{pool: pool}
+    {:ok, state}
+  end
+
+  def handle_call({:attempt, fun, opts}, from, state) do
+    case State.start_worker(state, fun, opts, from) do
+      {:ok, state} ->
+        {:noreply, state}
+
+      {:reject, exception} ->
+        {:reply, {:reject, exception}, state}
+    end
+  end
+
+  def handle_info({ref, result}, state) when is_reference(ref) do
+    state = State.result_received(state, ref, result)
+    {:noreply, state}
+  end
+
+  def handle_info({:DOWN, ref, :process, pid, :normal}, state)
+      when is_reference(ref) and is_pid(pid) do
+    {:noreply, state}
+  end
+
+  def handle_info(other, state) do
+    require Logger
+    Logger.error("#{__MODULE__} received unexpected message #{inspect(other)}")
+    {:noreply, state}
```

Comment:
This is an idiom that I've seen around other codebases (you can see an example of this in the official documentation https://elixir-lang.org/getting-started/mix-otp/genserver.html#the-need-for-monitoring).

As far as I've seen around the community the rationale is usually that some other process might attempt to send a debug or some other message that doesn't adhere to the protocol of this process, so it's better to log it as an error rather than crash and lose the state of the pool.

Maybe we should chat this one through to discuss the tradeoffs?

Reply:
Your reasoning makes sense here, as we wouldn't want to terminate the process if it happened to receive a random debug message (perhaps from a hypothetical profiling tool).